### ISSUE TITLE
A tree-sitter backend

### ISSUE THREAD

**Issue by wenkokke** [BNFC/bnfc#394]
If we create a [tree-sitter](https://tree-sitter.github.io/tree-sitter/) backend, we could get basic editor support for all languages using BNFC with almost no work.

Tree-sitter grammars are the defacto way of implementing highlighting in [Atom](https://flight-manual.atom.io/hacking-atom/sections/creating-a-grammar/), and there are packages which use tree-sitter grammars to provide highlighting in [VSCode](https://marketplace.visualstudio.com/items?itemName=evgeniypeshkov.syntax-highlighter), [neovim](https://github.com/nvim-treesitter), and [emacs](https://github.com/emacs-tree-sitter/elisp-tree-sitter). There are also [bindings to use a tree-sitter grammar from Java provided by JetBrains](https://github.com/JetBrains/jsitter), which would help with integration into the Jetbrains editor ecosystem.

There are bindings for tree-sitter are in various languages, including Haskell, JavaScript (both Node.js and Wasm), OCaml, Python, Ruby, and Rust.

Compiling a BNF grammar to tree-sitter should be fairly straightforward, and the only major hurdle I foresee would be to implement support for layouts, which would require some custom C code. For an example of how to implement this, one could look at the grammars for [Agda](https://github.com/tree-sitter/tree-sitter-agda), [Haskell](https://github.com/tree-sitter/tree-sitter-haskell), [Python](https://github.com/tree-sitter/tree-sitter-python), or any other language with layout rules.

**Comment by wenkokke** [comment]
Perhaps @banacorn could offer their advice, since they wrote the parser for Agda?￼￼

**Comment by andreasabel** [comment]
> > I wonder whether this could be factored into BNFC -> (E)BNF -> tree-sitter
> 
> Wouldn’t this lead to problems supporting layouts?

Yes, this wouldn't support layout, so maybe it is not worth looking into it, unless BNFC/-layout -> (E)BNF has its own interest.

**Comment by chaserhkj** [comment]
Hi, I have implemented a preliminary tree-sitter backend in #471. If any of the correspondents on this thread are interested, please help me test it out, thank you!
